Sources/SymbolKit/UnifiedSymbolGraph/UnifiedSymbolGraph.swift

```diff
@@ -160,6 +160,190 @@ extension UnifiedSymbolGraph {
+        // at index 2. In this case we need to recalculate the overload indices based on an ordering
+        // that takes all of the overloaded symbols into account.
+        //
+        // Finally, in situation 3 the relationship graph looks like this:
+        //
+        // ┌───────────────────────────────────────────┐
+        // │                 overloadA (1) ─┬─► groupA │
+        // │ groupB ◄─┬─ (1) overloadB (2) ─┘          │
+        // │          └─ (2) overloadC                 │
+        // └───────────────────────────────────────────┘
+        //
+        // In addition to overloadB having different indices per graph, the unified relationships
+        // are also inconsistent: there are two overload groups that refer to different subsets of
+        // symbols! In this case, we need to recalculate the overload indices like in situation 2
+        // above, but we also need to determine which overload group to keep as the "canonical" one.
+        // Since overload group symbols are by definition going to be generated by SymbolKit, when
+        // they're introduced into a unified symbol graph we can remove any that don't describe a
+        // "unified overload group" in this way, along with any relationships that point to them.
+        //
+        // In all three situations, we can generalize the solution with the following steps:
+        //
+        // 1. For each overload group, check the constituent symbols to see if they're part of
+        //    another overload group. Collect the symbols and groups as you go.
+        //    a. If you find other overload groups, recursively check for other overload groups
+        //       until no new groups are found.
+        // 2. Sort the collected symbols with the same comparator as
+        //    `SymbolGraph.createOverloadGroupSymbols()`.
+        // 3. Find the overload group symbol matching the first overload in the resulting order, and
+        //    ensure that it contains `overloadOf` relationships to all the overloaded symbols.
+        // 4. For each symbol, create a new OverloadData instance referencing the unified overload
+        //    group and the computed index in the sort order. Save it in the symbol's `unifiedMixins`.
+        // 5. If there were other overload group symbols, remove them from the symbol graph, along
+        //    with any relationships that referenced them.
+
+        /// A mapping of overload group symbols to their individual overloads.
+        var overloadGroups: [String: Set<String>] = [:]
+        /// A mapping of overloaded symbols to the overload groups they are part of.
+        var overloadGroupsBySymbol: [String: Set<String>] = [:]
+
+        for relationship in self.relationshipsByLanguage.flatMap(\.value)
+            where relationship.kind == SymbolGraph.Relationship.Kind.overloadOf
+        {
+            overloadGroups[relationship.target, default: []].insert(relationship.source)
```

Comment:
It might be worth adding an assertion here to check that each target, i.e. each overloadGroup, uses `SymbolGraph.Symbol.overloadGroupIdentifierSuffix`. This would also give the reader a clue of why the assert below should always be true: `siblingOverloadGroups.contains(computedOverloadGroup)`

Reply:
The `siblingOverloadGroups.contains` assertion doesn't have anything to do with the fact that the identifiers should reflect being an overload group. It's making sure that the unified graph view of the overloads (and its sorting method) reflected at least one of the constituent symbol graphs' view (and sorting) of the overloaded symbols. If the unified graph computes a different overload group than any of the individual graphs, that reflects a bug in the code. (One that i ran into several times while testing... 🥴)

Reply:
Got it thank you.